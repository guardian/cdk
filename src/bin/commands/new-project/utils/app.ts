import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  stages: string[];
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    const { comment, outputFile, imports, appName, stack, outputDir, stages } = this.config;

    this.code.openFile(outputFile);
    if (comment) {
      this.code.line(comment);
      this.code.line();
    }

    imports.render(this.code);

    this.code.line("const app = new App();");

    stages.forEach((stage) => {
      this.code.line(
        `new ${appName.pascal}(app, "${appName.pascal}-${stage}", { stack: "${stack.kebab}", stage: "${stage}" });`
      );
    });

    this.code.closeFile(outputFile);
    await this.code.save(outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
