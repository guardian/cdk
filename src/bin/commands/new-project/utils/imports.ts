import type { CodeMaker } from "codemaker";
import type { Name } from "./utils";

interface Import {
  types: string[];
  components: string[];
  basic?: boolean;
}

export class Imports {
  private readonly imports: Record<string, Import>;

  private constructor(imports: Record<string, Import>) {
    this.imports = imports;
  }

  public static newStackImports(addCfnIncludeImports: boolean): Imports {
    return new Imports({
      ...(addCfnIncludeImports && {
        path: {
          types: [],
          components: ["join"],
        },
        "aws-cdk-lib/cloudformation-include": {
          types: [],
          components: ["CfnInclude"],
        },
      }),
      "aws-cdk-lib": {
        types: ["App"],
        components: [],
      },
      "@guardian/cdk/lib/constructs/core": {
        types: ["GuStackProps"],
        components: ["GuStack"],
      },
    });
  }

  public static newAppImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "source-map-support/register": {
        basic: true,
        types: [],
        components: [],
      },
      [`../lib/${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  public static newTestImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "aws-cdk-lib/assertions": {
        types: [],
        components: ["Template"],
      },
      [`./${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  render(code: CodeMaker): void {
    Object.entries(this.imports)
      // Render "basic" imports before any others, still in alphabetical order
      // Render relative imports after absolute imports
      .sort(([aKey, aImports], [bKey, bImports]) => {
        if (aImports.basic && !bImports.basic) {
          return -1;
        } else if (bImports.basic && !aImports.basic) {
          return 1;
        } else if (aKey.startsWith(".") && !bKey.startsWith(".")) {
          return 1;
        } else if (bKey.startsWith(".") && !aKey.startsWith(".")) {
          return -1;
        } else {
          return aKey.localeCompare(bKey);
        }
      })
      .forEach(([lib, imports]) => {
        imports.basic && code.line(`import "${lib}";`);

        imports.types.length && code.line(`import type { ${imports.types.sort().join(", ")} } from "${lib}";`);

        imports.components.length && code.line(`import { ${imports.components.sort().join(", ")} } from "${lib}";`);
      });
    code.line();
  }
}
