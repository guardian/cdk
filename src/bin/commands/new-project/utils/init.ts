import { copyFileSync, existsSync, lstatSync, mkdirSync, readdirSync, writeFileSync } from "fs";
import { join } from "path";
import { LibraryInfo } from "../../../../constants";

export interface InitConfig {
  outputDir: string;
}

// TODO: Add yarn or npm flag
// TODO: Add project name flag
export class ProjectBuilder {
  templateDir = `${__dirname}/../template`;
  config: InitConfig;

  static validateConfig = (config: InitConfig): void => {
    const path = config.outputDir;
    if (existsSync(path) && readdirSync(path).length > 0) {
      throw new Error(`Directory ${path} is not empty`);
    }
  };

  constructor(config: InitConfig) {
    this.config = config;
  }

  buildDirectory(): void {
    ProjectBuilder.validateConfig(this.config);

    if (!existsSync(this.config.outputDir)) {
      console.log(`Creating ${this.config.outputDir}`);
      mkdirSync(this.config.outputDir);
    }

    console.log("Creating package.json");
    createPackageJson(this.config.outputDir);

    console.log("Copying template files");
    // TODO: Replace any params in files with .template extensions
    this.copyFiles(this.templateDir, this.config.outputDir);

    console.log("Success!");
  }

  copyFiles(sourcePath: string, targetPath: string): void {
    for (const file of readdirSync(sourcePath)) {
      const path = join(sourcePath, file);

      if (path.endsWith(".ignore")) {
        continue;
      } else if (lstatSync(path).isDirectory()) {
        const nestedTargetPath = join(targetPath, file);
        if (!existsSync(nestedTargetPath)) {
          mkdirSync(nestedTargetPath);
        }
        this.copyFiles(path, nestedTargetPath);
      } else if (path.endsWith(".template")) {
        copyFileSync(path, join(targetPath, file.replace(".template", "")));
      } else {
        copyFileSync(path, join(targetPath, file));
      }
    }
  }
}

export const buildDirectory = (config: InitConfig): void => {
  const builder = new ProjectBuilder(config);
  builder.buildDirectory();
};

function createPackageJson(outputDirectory: string): void {
  const { NODE_ENV, CI } = process.env;
  const isTest = NODE_ENV?.toUpperCase() === "TEST" || CI?.toUpperCase() === "TRUE";

  const contents = {
    name: "cdk",
    version: "0.0.0",
    private: true,
    bin: {
      cdk: "bin/cdk.js",
    },
    scripts: {
      build: "tsc",
      watch: "tsc -w",
      test: "jest",
      "test:dev": "jest --watch",
      format: 'prettier --write "{lib,bin}/**/*.ts"',
      cdk: "cdk",
      lint: "eslint lib/** bin/** --ext .ts --no-error-on-unmatched-pattern",
      synth: "cdk synth --path-metadata false --version-reporting false",
      diff: "cdk diff --path-metadata false --version-reporting false",
    },
    devDependencies: {
      "@aws-cdk/assert": LibraryInfo.AWS_CDK_VERSION,
      "@guardian/eslint-config-typescript": "^0.7.0",
      "@types/jest": "^27.0.2",
      "@types/node": "16.11.7",
      "aws-cdk": LibraryInfo.AWS_CDK_VERSION,
      eslint: "^7.32.0",
      jest: "^27.3.1",
      prettier: "^2.4.1",
      "ts-jest": "^27.0.7",
      "ts-node": "^10.4.0",
      typescript: "~4.4.3",
    },
    dependencies: {
      "@aws-cdk/cloudformation-include": LibraryInfo.AWS_CDK_VERSION,
      "@aws-cdk/core": LibraryInfo.AWS_CDK_VERSION,

      /*
      Do not add `@guardian/cdk` to the generated `package.json` file when in TEST as we'll `npm link` it instead.
      See https://docs.npmjs.com/cli/v8/commands/npm-link#caveat

      TODO remove this once the `new` command allows opting out of automatic dependency installation
       */
      ...(!isTest && { "@guardian/cdk": LibraryInfo.VERSION }),

      "source-map-support": "^0.5.20",
    },
  };
  writeFileSync(`${outputDirectory}/package.json`, JSON.stringify(contents, null, 2));
}
