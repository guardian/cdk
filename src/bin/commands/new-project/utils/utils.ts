import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
