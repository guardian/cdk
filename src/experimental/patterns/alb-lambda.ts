import type {NoMonitoring} from "../../constructs/cloudwatch";
import type {GuStack} from "../../constructs/core";
import type {GuFunctionProps} from "../../constructs/lambda";
import {GuLambdaFunction} from "../../constructs/lambda";
import {
  GuApplicationLoadBalancer,
  GuApplicationTargetGroup,
  GuHttpsApplicationListener
} from "../../constructs/loadbalancing";
import {AccessScope} from "../../constants";
import {AccessLoggingProps, Alarms} from "../../patterns/ec2-app";
import {LambdaTarget} from "aws-cdk-lib/aws-elasticloadbalancingv2-targets";
import {GuCertificate} from "../../constructs/acm";
import type {GuDomainName} from "../../types";
import {GuLoadBalancingComponents} from "../../patterns/load-balancer/load-balancer";

export interface GuAlbLambdaProps extends Omit<GuFunctionProps, "errorPercentageMonitoring"> {
  /**
   *
   * Enable and configure alarms.
   *
   * If your team do not use CloudWatch, it's possible to opt-out with the following configuration:
   * ```typescript
   *  monitoringConfiguration: { noMonitoring: true }
   * ```
   */
  monitoringConfiguration: Alarms | NoMonitoring;

  accessLogging?: AccessLoggingProps;

  certificateProps?: GuDomainName;
}

/**
 * A pattern to create a Lambda triggered by a load balancer
 * @see ttps://docs.aws.amazon.com/lambda/latest/dg/services-alb.html
 *
 * This pattern should be used if you intend to serve all traffic via a single Lambda
 * (for example, if your Lambda uses an application framework, like https://github.com/vendia/serverless-express).
 *
 * For all configuration options, see [[`GuAlbLambdaProps`]].
 *
 * Example usage:
 *
 * ```typescript
 * new GuAlbLambda(stack, "lambda", {
 *       fileName: "my-app.zip",
 *       handler: "handler.ts",
 *       runtime: Runtime.NODEJS_12_X,
 *       monitoringConfiguration: noMonitoring,
 *       app: "testing",
 *       vpc: vpc,
 *       vpcSubnets: {
 *         subnets: vpc.privateSubnets
 *       }
 *     });
 * ```
 */
export class GuAlbLambda extends GuLambdaFunction {
  public readonly certificate?: GuCertificate;
  public readonly loadBalancer: GuApplicationLoadBalancer;
  public readonly listener: GuHttpsApplicationListener;
  public readonly targetGroup: GuApplicationTargetGroup;

  constructor(scope: GuStack, id: string, props: GuAlbLambdaProps) {
    super(scope, id, {
      ...props,
    });

    if (!props.vpc) {
      throw new Error("VPC must be defined for ALB lambdas")
    }

    const targetGroup = new GuApplicationTargetGroup(scope, "TargetGroup", {
      targetGroupType: "lambda",
      app: props.app,
      vpc: props.vpc,
      targets: [new LambdaTarget(this)]
    })


    const loadBalancer = new GuLoadBalancingComponents(scope, {
      ...props,
      access: { scope: AccessScope.PUBLIC },
      targetGroup: targetGroup
    })

    this.certificate = loadBalancer.certificate;
    this.loadBalancer = loadBalancer.loadBalancer;
    this.listener = loadBalancer.listener;
    this.targetGroup = loadBalancer.targetGroup;
  }
}
