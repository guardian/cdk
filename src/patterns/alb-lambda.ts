import type {NoMonitoring} from "../constructs/cloudwatch";
import type {GuStack} from "../constructs/core";
import type {GuFunctionProps} from "../constructs/lambda";
import {GuLambdaFunction} from "../constructs/lambda";
import {
  GuApplicationLoadBalancer,
  GuApplicationTargetGroup,
  GuHttpsApplicationListener
} from "../constructs/loadbalancing";
import {AccessScope} from "../constants";
import {AccessLoggingProps, Alarms} from "./ec2-app";
import {LambdaTarget} from "aws-cdk-lib/aws-elasticloadbalancingv2-targets";
import {GuCertificate} from "../constructs/acm";
import type {GuDomainName} from "../types";
import {GuLoadBalancingComponents} from "./load-balancer/load-balancer";

export interface GuAlbLambdaProps extends Omit<GuFunctionProps, "errorPercentageMonitoring"> {
  /**
   * Alarm configuration for your API. For more details, see [[`ApiGatewayAlarms`]].
   *
   * If your team do not use CloudWatch, it's possible to opt-out with the following configuration:
   * ```typescript
   *  monitoringConfiguration: { noMonitoring: true }
   * ```
   */
  monitoringConfiguration: Alarms | NoMonitoring;

  accessLogging?: AccessLoggingProps;

  certificateProps?: GuDomainName;
}

interface GuAlbLambdaPropsTarget extends GuAlbLambdaProps {
  // target: IApplicationLoadBalancerTarget;

}

/**
 * A pattern to create a Lambda triggered by API Gateway
 * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/getting-started-with-lambda-integration.html
 *
 * This pattern should be used if you intend to serve all traffic via a single Lambda
 * (for example, if your Lambda uses an application framework, like https://github.com/vendia/serverless-express).
 * If you need to configure path-based routing to serve different requests with different
 * Lambdas, use the [[`GuApiGatewayWithLambdaByPath`]] pattern instead.
 *
 * For all configuration options, see [[`GuApiLambdaProps`]].
 *
 * Example usage:
 *
 * ```typescript
 * new GuApiLambda(stack, "my-lambda", {
 *   fileName: "my-app.zip",
 *   handler: "handler.ts",
 *   runtime: Runtime.NODEJS_14_X,
 *   monitoringConfiguration: {
 *     http5xxAlarm: { tolerated5xxPercentage: 5 },
 *     snsTopicName: "alerts-topic",
 *   },
 *   app: "my-app",
 *   api: {
 *     id: "my-api",
 *     description: "...",
 *   },
 * });
 * ```
 */
export class GuAlbLambda extends GuLambdaFunction {
  public readonly certificate?: GuCertificate;
  public readonly loadBalancer: GuApplicationLoadBalancer;
  public readonly listener: GuHttpsApplicationListener;
  public readonly targetGroup: GuApplicationTargetGroup;

  constructor(scope: GuStack, id: string, props: GuAlbLambdaPropsTarget) {
    super(scope, id, {
      ...props,
    });

    if (!props.vpc) {
      throw new Error("VPC must be defined for ALB lambdas")
    }

    const targetGroup = new GuApplicationTargetGroup(scope, "TargetGroup", {
      app: props.app,
      vpc: props.vpc,
      targets: [new LambdaTarget(this)]
    })


    const loadBalancer = new GuLoadBalancingComponents(scope, {
      ...props,
      access: { scope: AccessScope.PUBLIC },
      targetGroup: targetGroup
    })

    this.certificate = loadBalancer.certificate;
    this.loadBalancer = loadBalancer.loadBalancer;
    this.listener = loadBalancer.listener;
    this.targetGroup = loadBalancer.targetGroup;
  }
}
