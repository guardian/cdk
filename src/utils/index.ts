export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
