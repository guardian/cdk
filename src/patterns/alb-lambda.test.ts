import { Template } from "aws-cdk-lib/assertions";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import type { NoMonitoring } from "../constructs/cloudwatch";
import { simpleGuStackForTesting } from "../utils/test";
import {GuAlbLambda} from "./alb-lambda";
import {Vpc} from "aws-cdk-lib/aws-ec2";
import {Stack} from "aws-cdk-lib";

describe("The GuAlbLambda pattern", () => {
  const vpc = Vpc.fromVpcAttributes(new Stack(), "VPC", {
    vpcId: "test",
    availabilityZones: [""],
    publicSubnetIds: [""],
    privateSubnetIds: ["test1"]
  });

  it("should create the correct resources with minimal config", () => {
    const stack = simpleGuStackForTesting();
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    new GuAlbLambda(stack, "lambda", {
      fileName: "my-app.zip",
      handler: "handler.ts",
      runtime: Runtime.NODEJS_12_X,
      monitoringConfiguration: noMonitoring,
      app: "testing",
      vpc: vpc,
      vpcSubnets: {
        subnets: vpc.privateSubnets
      }
    });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  // it("should create an alarm if monitoring configuration is provided", () => {
  //   const stack = simpleGuStackForTesting();
  //   new GuAlbLambda(stack, "lambda", {
  //     fileName: "my-app.zip",
  //     handler: "handler.ts",
  //     runtime: Runtime.NODEJS_12_X,
  //     monitoringConfiguration: {
  //       http5xxAlarm: { tolerated5xxPercentage: 5 },
  //       snsTopicName: "alerts-topic",
  //     },
  //     app: "testing",
  //     api: {
  //       id: "api",
  //       description: "this is a test",
  //     },
  //   });
  //   Template.fromStack(stack).resourceCountIs("AWS::CloudWatch::Alarm", 1);
  // });

  // it("should allow us to link a domain name to a LambdaRestApi", () => {
  //   const stack = simpleGuStackForTesting();
  //   const noMonitoring: NoMonitoring = { noMonitoring: true };
  //   const apiLambda = new GuApiLambda(stack, "lambda", {
  //     fileName: "my-app.zip",
  //     handler: "handler.ts",
  //     runtime: Runtime.NODEJS_12_X,
  //     monitoringConfiguration: noMonitoring,
  //     app: "testing",
  //     api: {
  //       id: "api",
  //       description: "this is a test",
  //     },
  //   });
  //
  //   const domain = apiLambda.api.addDomainName("domain", {
  //     domainName: "code.theguardian.com",
  //     certificate: new GuCertificate(stack, {
  //       app: "testing",
  //       domainName: "code.theguardian.com",
  //       hostedZoneId: "id123",
  //     }),
  //   });
  //
  //   new GuCname(stack, "DNS", {
  //     app: "testing",
  //     ttl: Duration.days(1),
  //     domainName: "code.theguardian.com",
  //     resourceRecord: domain.domainNameAliasDomainName,
  //   });
  //   expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  // });
  //
  // it("should route requests to $LATEST (i.e. an unpublished version) if the enableVersioning prop is unset", () => {
  //   const stack = simpleGuStackForTesting();
  //   const noMonitoring: NoMonitoring = { noMonitoring: true };
  //   new GuApiLambda(stack, "lambda", {
  //     fileName: "my-app.zip",
  //     handler: "handler.ts",
  //     runtime: Runtime.NODEJS_12_X,
  //     monitoringConfiguration: noMonitoring,
  //     app: "testing",
  //     api: {
  //       id: "api",
  //       description: "this is a test",
  //     },
  //   });
  //   Template.fromStack(stack).hasResourceProperties("AWS::ApiGateway::Method", {
  //     Integration: {
  //       Uri: {
  //         "Fn::Join": [
  //           "",
  //           [
  //             "arn:",
  //             {
  //               Ref: "AWS::Partition",
  //             },
  //             ":apigateway:",
  //             {
  //               Ref: "AWS::Region",
  //             },
  //             ":lambda:path/2015-03-31/functions/",
  //             {
  //               "Fn::GetAtt": ["lambda8B5974B5", "Arn"],
  //             },
  //             "/invocations",
  //           ],
  //         ],
  //       },
  //     },
  //   });
  // });
  //
  // it("should route requests to an alias if the enableVersioning prop is set to true", () => {
  //   const stack = simpleGuStackForTesting();
  //   const noMonitoring: NoMonitoring = { noMonitoring: true };
  //   new GuApiLambda(stack, "lambda", {
  //     enableVersioning: true,
  //     fileName: "my-app.zip",
  //     handler: "handler.ts",
  //     runtime: Runtime.NODEJS_12_X,
  //     monitoringConfiguration: noMonitoring,
  //     app: "testing",
  //     api: {
  //       id: "api",
  //       description: "this is a test",
  //     },
  //   });
  //   Template.fromStack(stack).hasResourceProperties("AWS::ApiGateway::Method", {
  //     Integration: {
  //       Uri: {
  //         "Fn::Join": [
  //           "",
  //           [
  //             "arn:",
  //             {
  //               Ref: "AWS::Partition",
  //             },
  //             ":apigateway:",
  //             {
  //               Ref: "AWS::Region",
  //             },
  //             ":lambda:path/2015-03-31/functions/",
  //             {
  //               Ref: "lambdaAliasForLambda426FBB83", // This is the important difference when compared to the test above
  //             },
  //             "/invocations",
  //           ],
  //         ],
  //       },
  //     },
  //   });
  // });
});
