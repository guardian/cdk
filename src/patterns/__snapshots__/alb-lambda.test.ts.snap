// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuAlbLambda pattern should create the correct resources with minimal config 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStack",
      "GuDistributionBucketParameter",
      "GuAlbLambda",
      "GuSubnetListParameter",
      "GuSubnetListParameter",
      "GuApplicationLoadBalancer",
      "GuApplicationTargetGroup",
      "GuHttpsApplicationListener",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "LoadBalancerTestingDnsName": {
      "Description": "DNS entry for LoadBalancerTesting",
      "Value": {
        "Fn::GetAtt": [
          "LoadBalancerTesting006FFF6B",
          "DNSName",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "testingPrivateSubnets": {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "testingPublicSubnets": {
      "Default": "/account/vpc/primary/subnets/public",
      "Description": "A list of public subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
  },
  "Resources": {
    "ListenerTesting34F16D51": {
      "Properties": {
        "DefaultActions": [
          {
            "TargetGroupArn": {
              "Ref": "TargetGroupTesting29B71ABC",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "LoadBalancerTesting006FFF6B",
        },
        "Port": 8080,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LoadBalancerTesting006FFF6B": {
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "LoadBalancerTestingSecurityGroup8250E10B",
              "GroupId",
            ],
          },
        ],
        "Subnets": {
          "Ref": "testingPublicSubnets",
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "testing",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "LoadBalancerTestingSecurityGroup8250E10B": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB TestLoadBalancerTesting6B05289C",
        "SecurityGroupEgress": [
          {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "testing",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": "test",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TargetGroupTesting29B71ABC": {
      "DependsOn": [
        "lambdaInvoke2UTWxhlfyqbT5FTn5jvgbLgjFfJwzswGk55DU1HY67C4BB0E",
      ],
      "Properties": {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/healthcheck",
        "HealthCheckProtocol": "HTTP",
        "HealthCheckTimeoutSeconds": 5,
        "HealthyThresholdCount": 5,
        "Tags": [
          {
            "Key": "App",
            "Value": "testing",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TargetGroupAttributes": [
          {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "30",
          },
        ],
        "TargetType": "lambda",
        "Targets": [
          {
            "Id": {
              "Fn::GetAtt": [
                "lambda8B5974B5",
                "Arn",
              ],
            },
          },
        ],
        "UnhealthyThresholdCount": 2,
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "lambda8B5974B5": {
      "DependsOn": [
        "lambdaServiceRoleDefaultPolicyBF6FA5E7",
        "lambdaServiceRole494E4CA6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "test-stack/TEST/testing/my-app.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "testing",
            "STACK": "test-stack",
            "STAGE": "TEST",
          },
        },
        "Handler": "handler.ts",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "lambdaServiceRole494E4CA6",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "testing",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
        "VpcConfig": {
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "lambdaSecurityGroup68370914",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": [
            "test1",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaInvoke2UTWxhlfyqbT5FTn5jvgbLgjFfJwzswGk55DU1HY67C4BB0E": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "lambda8B5974B5",
            "Arn",
          ],
        },
        "Principal": "elasticloadbalancing.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "lambdaSecurityGroup68370914": {
      "Properties": {
        "GroupDescription": "Automatic security group for Lambda Function Testlambda2DC975DC",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "testing",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": "test",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "lambdaServiceRole494E4CA6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "testing",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaServiceRoleDefaultPolicyBF6FA5E7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/test-stack/TEST/testing/my-app.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/test-stack/testing",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/test-stack/testing/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaServiceRoleDefaultPolicyBF6FA5E7",
        "Roles": [
          {
            "Ref": "lambdaServiceRole494E4CA6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
